### ISSUE TITLE
convert old slack_relay to a service

### ISSUE THREAD

**Issue by laroque** [project8/dragonfly#56]
There was previously a dripline subcommand slack_relay which was never ported to dragonfly. It should actually be a service (Gogol under the current lexicon, though that should get changed a la dripline issue 162). At first pass this service would just replace the slack handler (eliminating the need to distribute slack tokens and support slack in every service).
## Motiviation
1. Allows dripline services in other languages (c++ in particular) to post to the slack channels without needing their own slack dependencies. This was the original motivation.
2. Allows silencing of slack noise (just stop one service) without disrupting the running control systems.
3. Provides a single point for implementing logic. We could do whatever we want but the obvious things would be:
   - collect messages and publish only at some fixed interval, preventing possible future API limit issues
   - could implement a feature to silence a particular source
   - could suppress duplicate messages within some short time interval
   - could more easily split messages between #p8_[critical, alerts, notices], eliminating operator phone notifications for things that are non-urgent,

**Comment by wcpettus** [comment]
With the phasing out of slack_handler, does that mean the entire dragonfly/status_log_handlers set will be retired?  Or will they be resurrected in a different way?

**Comment by laroque** [comment]
dragonfly/status_log_handlers will probably still need to exist. It turns out that the default python logging.StreamHandler sends everything to stderr. For supervisord it would be much nicer to split between stdout and stderr which would be pretty easy with a couple of only slightly modified handlers.

Anything that talks to an external source (slack, twitter if we ever wanted to use it again, emailed or sms which we don't have now but could) should probably be a central service for the same motivations above.

**Comment by wcpettus** [comment]
If this issue gets fixed, tag #42 to check for similar resolution.

**Comment by laroque** [comment]
Some log handlers (slack in particular) should support some sort of digest form so that we don't exceed the API limits (and also to reduce the rate of alerts).

The logging module already supports this structure with the MemoryHandler. We could implement a two layer system where a MemoryHandler uses the existing slack logger as its target. With the following behaviors:

in emit(), if shouldFlush() returns false, we should do something to ensure that flush is called reasonably soon, probably using the AMQP backend to schedule as is done with the data logging in the scheduler
in flush(), a timestamp of the current flush should be recorded
shouldFlush() should return true if time since the last flush is greater than some limit.

**Comment by guiguem** [comment]
What is the status of this SlackRelay? Has it been tested? What features are missing?

**Comment by laroque** [comment]
There are two parts. The first is well defined and somewhat tested but will require time and thought to ensure we cover the edge cases and leave room for enhancements we may want (like aggregating messages into a single post)

1) There was an old slack relay which is most of an implementations, but like the ESR it was implemented as a subcommand. It should be converted to a service. The dripline standard already has descriptions of the messages that should be sent/received by that service.

2) Modify service behavior so that messages get sent. I'd say that we should add a utility method, parallel to get() or set(), which sends a status broadcast. We then either go through the codebase and convert logger.critical() to be calls to this method, or we create a custom log handler that catches critical messages and sends dripline alerts of this type.

**Comment by guiguem** [comment]
![fullsizerender](https://cloud.githubusercontent.com/assets/16764756/20195289/c64bb996-a74a-11e6-9379-3c681257b2c2.jpg)
the slack handler already exists and needs to send alerts to the AMQP. the slack interface (gogol) needs to be adapt to as a service and send the alerts message to the slack api
There are more features to do also, but is this scheme correct?

**Comment by laroque** [comment]
Almost exactly. Right now there are multiple handlers (twitter, slack, whatever) because they make the api calls directly. Under the new scheme this is only one custom handler**, the amqp handler, which sends alert messages. Those go to the slack interface exactly as you describe. If we want to have twitter again, we'd add a new Gogol, which takes the alerts and posts using the twitter API (or similar for email or sms or anything we decide to add), but don't need any new handlers.

**that statement isn't 100% correct. There is a handler which writes console output and we may want to customize that to split STDOUT and STDERR (see #48) in a non-default way. But the only handlers would be local (console-like) or AMQP.

**Comment by guiguem** [comment]
ok, so no emit_via_api between slack_handler (renamed AMQPhandler) and slack, but every message we want to go to slack should go through the AMQP server. Great! I see the picture now, thank you.

**Comment by guiguem** [comment]
We have a problem here which comes from dragonfly itself: the AMQP_handler cannot connect to the broker (and therefore cannot be created as an instance), since the handler is created during initialization of the Dripline parser and the broker name is parsed later.
We need to allow to create the handler instance right after the parsing of the broker:

**Comment by laroque** [comment]
I think I see the problem, the solution is in two steps.

1) Add a convenience function, service.send_status_alert()
2) Change the dragonfly executable so that it creates the AMQP_handler after it initializes the service. Pass the service to it on __init__ so that it can use service.send_status_alert().

3) <cleanup> Remove the twitter handler, slack handler, and other deprecated handlers, but let the console handlers continue to be created early as they already are, it is only the AMQP handler that needs to be configured later.

**Comment by guiguem** [comment]
I am currently using the service.send_alert method in the AMQP handler (see commit 50f4fc84e9da7c6854d3a07f6863fc16b60d23d5).
I changed the order of creating the handlers (other than the default one): the other handlers are now created in the parse_args of dragonfly.
If it is a AMQP handler we are about to create, I pass the broker name to the handler now via the kwargs.
I am going to depreciate the twitter handler.

**Comment by guiguem** [comment]
Ok, I have a working version of the slack service. The AMQP handler now sends every critical log  to the AMQP service via alerts.
A SlackInterface Gogol service will get these messages and send a message on Slack.
If messages are sent too often, there are silenced after a warning. Warnings do not happen every time (one can select the delta t between 2 warnings).
Currently, one problem remains: critical logs and standard alerts are not associated with the same service name (the SpimeScapelike rsa_daq will be associated with critical logs and the Endpoint like rsa_daq_interface will be associated to the alerts).
Might need a workaround, or this can wait since almost all the elements we wanted are here.
@nsoblath the generic dripline Slack username has currently a special treatment: they are prime speakers and are allowed to talk how often they want. At some point, do we want to migrate the interaction of hornet with Slack to this new service? hornet will send alerts, and they will be interpreted. hornet could be a prime speaker too.

**Comment by laroque** [comment]
I'm not sure what this means: "critical logs and standard alerts are not associated with the same service name"? The format is described here: https://github.com/project8/dripline/wiki/Use-Protocol#alerts. (which should match what the old "not yet ported" subcommand did. The routing key's third element should indicate where the alert is coming from.

Also, how is this "sent too often" defined? Is it on a per-source basis, or all messages? The later case is almost certain to come back to bite us when more than one service tries to send an alert at more or less the same time. Collecting messages and posting them in a digest form seems reasonable though. Also is the silencing permanent? That also seems likely to come back to bite us hard.

**Comment by guiguem** [comment]
The routing key's third element indeed indicates where the message comes from, but depending on the origin of the message (logger.critical or directly sending an alert), this element is different: for example, for the rsa_daq
```yaml
name: rsa_daq
module: Spimescape
broker: myrna.p8
endpoints:
  - name: rsa_daq_interface
    module: RSAAcquisitionInterface
```
a critical will have this third element as rsa_daq (and therefore Slack will receive a message from rsa_daq) and if the rsa_daq_interface wants to send an alert, the 3rd element of routine key will be rsa_daq_interface: they will not appear with the same username even if they are kind of coming from the same service... that is a cosmetic detail, but I wanted to mention it.

The 'sent too often' trigger is defined on per-source basis. The config file has two items which are the number of allowed messages per speaking time and the speaking time. A dictionary is but inside the SlackInterface class and keeps track of who sent a message and when. If something sends a message too often, it is muted until the number of messages published on slack from this thing gets below the limit.
From time to time (the time being defined in the config file too, but something like one hour), a warning is send to remind us that something tries to complain.

I tried to design the thing such as these possible problems are well taken into account.
My guess is that I will prepare a preliminary version of this, with some appropriate parameters and try to have it running with the new release; we will then be able to see what we actually need as a muting time or warning period.

**Comment by laroque** [comment]
That sounds great.

**Comment by guiguem** [comment]
A first version of the SlackInterface and the AMQP Handler have been released with v1.3.0. The main features should be working, we might need some improvements as the time goes. We should open a new issue for desired extra enhancements.
